File: vscode-extension/src/credentialConfig.test.ts
import { SECRET_KEY_ANTHROPIC_API_KEY, SecretsHost } from "./apiKeyCommand";
import {
  buildProxyEnv,
  clearCredentials,
  ensureCredentials,
  GatewayCredentials,
  MODE_LABEL_DIRECT,
  MODE_LABEL_GATEWAY,
  runSetCredentials,
  SECRET_KEY_CREDENTIAL_MODE,
  SECRET_KEY_GATEWAY_AUTH_TOKEN,
  SECRET_KEY_GATEWAY_BASE_URL,
  SECRET_KEY_GATEWAY_SKIP_TLS_VERIFY,
} from "./credentialConfig";

/** In-memory fake, same role as apiKeyCommand.test.ts's makeFakeHost --
 * `store` doubles as the "secure storage" for persistence-across-restart
 * assertions (a fresh FakeSecretsHost pointed at the SAME store object
 * simulates "the IDE restarted, SecretStorage/PasswordSafe is still
 * there"). */
function makeFakeHost(options: {
  store?: Record<string, string>;
  choice?: string;
  gatewayToken?: string;
  gatewayUrl?: string;
  skipTlsVerify?: boolean;
  apiKey?: string;
} = {}): SecretsHost & { store: Record<string, string>; infoMessages: string[]; errorMessages: string[] } {
  const store = options.store ?? {};
  const infoMessages: string[] = [];
  const errorMessages: string[] = [];
  return {
    store,
    infoMessages,
    errorMessages,
    async getSecret(key) {
      return store[key];
    },
    async storeSecret(key, value) {
      store[key] = value;
    },
    async deleteSecret(key) {
      delete store[key];
    },
    async promptForApiKey() {
      return options.apiKey;
    },
    async promptForChoice() {
      return options.choice;
    },
    async promptForText(promptText) {
      if (promptText.includes("gateway URL") || promptText.includes("base URL")) {
        return options.gatewayUrl;
      }
      return options.gatewayToken;
    },
    async confirmDangerousToggle() {
      return options.skipTlsVerify ?? false;
    },
    showInfo(message) {
      infoMessages.push(message);
    },
    showError(message) {
      errorMessages.push(message);
    },
  };
}

describe("runSetCredentials", () => {
  it("cancel at the mode choice stores nothing", async () => {
    const host = makeFakeHost({ choice: undefined });
    const stored = await runSetCredentials(host);
    expect(stored).toBe(false);
    expect(Object.keys(host.store)).toHaveLength(0);
  });

  it("Direct mode stores the API key and mode flag", async () => {
    const host = makeFakeHost({ choice: MODE_LABEL_DIRECT, apiKey: "sk-ant-abc" });
    const stored = await runSetCredentials(host);
    expect(stored).toBe(true);
    expect(host.store[SECRET_KEY_ANTHROPIC_API_KEY]).toBe("sk-ant-abc");
    expect(host.store[SECRET_KEY_CREDENTIAL_MODE]).toBe("direct");
  });

  it("Org Gateway mode stores token, URL, mode flag, and TLS-skip=false by default", async () => {
    const host = makeFakeHost({
      choice: MODE_LABEL_GATEWAY,
      gatewayToken: "  gw-token-123  ",
      gatewayUrl: "https://api-eu1.aigateway.example.com",
    });
    const stored = await runSetCredentials(host);
    expect(stored).toBe(true);
    expect(host.store[SECRET_KEY_GATEWAY_AUTH_TOKEN]).toBe("gw-token-123");
    expect(host.store[SECRET_KEY_GATEWAY_BASE_URL]).toBe("https://api-eu1.aigateway.example.com");
    expect(host.store[SECRET_KEY_GATEWAY_SKIP_TLS_VERIFY]).toBe("false");
    expect(host.store[SECRET_KEY_CREDENTIAL_MODE]).toBe("gateway");
    expect(host.infoMessages[0]).not.toContain("gw-token-123");
  });

  it("Org Gateway mode with TLS-skip explicitly confirmed stores 'true' and warns in the confirmation", async () => {
    const host = makeFakeHost({
      choice: MODE_LABEL_GATEWAY,
      gatewayToken: "gw-token",
      gatewayUrl: "https://gw.example.com",
      skipTlsVerify: true,
    });
    await runSetCredentials(host);
    expect(host.store[SECRET_KEY_GATEWAY_SKIP_TLS_VERIFY]).toBe("true");
    expect(host.infoMessages[0]).toContain("DISABLED");
  });

  it("cancelling the token prompt in gateway mode stores nothing", async () => {
    const host = makeFakeHost({ choice: MODE_LABEL_GATEWAY, gatewayToken: undefined, gatewayUrl: "https://gw.example.com" });
    const stored = await runSetCredentials(host);
    expect(stored).toBe(false);
    expect(Object.keys(host.store)).toHaveLength(0);
  });

  it("cancelling the URL prompt in gateway mode stores nothing (not even the already-entered token)", async () => {
    const host = makeFakeHost({ choice: MODE_LABEL_GATEWAY, gatewayToken: "gw-token", gatewayUrl: undefined });
    const stored = await runSetCredentials(host);
    expect(stored).toBe(false);
    expect(Object.keys(host.store)).toHaveLength(0);
  });
});

describe("ensureCredentials", () => {
  it("resolves stored Direct-mode credentials without prompting", async () => {
    const host = makeFakeHost({
      store: { [SECRET_KEY_ANTHROPIC_API_KEY]: "sk-ant-existing", [SECRET_KEY_CREDENTIAL_MODE]: "direct" },
    });
    const creds = await ensureCredentials(host);
    expect(creds).toEqual({ mode: "direct", apiKey: "sk-ant-existing" });
  });

  it("resolves stored Org Gateway credentials without prompting", async () => {
    const host = makeFakeHost({
      store: {
        [SECRET_KEY_CREDENTIAL_MODE]: "gateway",
        [SECRET_KEY_GATEWAY_AUTH_TOKEN]: "gw-token",
        [SECRET_KEY_GATEWAY_BASE_URL]: "https://gw.example.com",
        [SECRET_KEY_GATEWAY_SKIP_TLS_VERIFY]: "true",
      },
    });
    const creds = await ensureCredentials(host);
    expect(creds).toEqual({ mode: "gateway", authToken: "gw-token", baseUrl: "https://gw.example.com", skipTlsVerify: true });
  });

  it("credential-mode choice persists across a simulated IDE restart (fresh host, same backing store)", async () => {
    const sharedStore: Record<string, string> = {};
    const firstSessionHost = makeFakeHost({
      store: sharedStore,
      choice: MODE_LABEL_GATEWAY,
      gatewayToken: "gw-token-persisted",
      gatewayUrl: "https://gw.example.com",
    });
    await runSetCredentials(firstSessionHost);

    // A brand-new host instance pointed at the same store simulates a
    // fresh extension activation after a restart -- no prompt functions
    // are wired to return anything, so if this had to prompt again the
    // test would fail (getSecret is the ONLY thing exercised).
    const secondSessionHost = makeFakeHost({ store: sharedStore });
    const creds = await ensureCredentials(secondSessionHost);
    expect(creds).toEqual({
      mode: "gateway",
      authToken: "gw-token-persisted",
      baseUrl: "https://gw.example.com",
      skipTlsVerify: false,
    });
  });

  it("falls through to re-prompting if mode says gateway but credentials are incomplete (never silently defaults to direct)", async () => {
    const host = makeFakeHost({
      store: { [SECRET_KEY_CREDENTIAL_MODE]: "gateway" }, // token/URL missing
      choice: MODE_LABEL_GATEWAY,
      gatewayToken: "recovered-token",
      gatewayUrl: "https://gw.example.com",
    });
    const creds = await ensureCredentials(host);
    expect(creds?.mode).toBe("gateway");
    expect((creds as GatewayCredentials).authToken).toBe("recovered-token");
  });

  it("returns undefined if the user cancels the prompt with nothing stored yet", async () => {
    const host = makeFakeHost({ choice: undefined });
    const creds = await ensureCredentials(host);
    expect(creds).toBeUndefined();
  });
});

describe("clearCredentials", () => {
  it("removes all credential keys for both modes", async () => {
    const host = makeFakeHost({
      store: {
        [SECRET_KEY_ANTHROPIC_API_KEY]: "sk-ant",
        [SECRET_KEY_GATEWAY_AUTH_TOKEN]: "gw-token",
        [SECRET_KEY_GATEWAY_BASE_URL]: "https://gw.example.com",
        [SECRET_KEY_GATEWAY_SKIP_TLS_VERIFY]: "true",
        [SECRET_KEY_CREDENTIAL_MODE]: "gateway",
      },
    });
    await clearCredentials(host);
    expect(Object.keys(host.store)).toHaveLength(0);
  });
});

describe("buildProxyEnv", () => {
  it("Direct mode: sets ANTHROPIC_API_KEY only, matching chat_proxy.py's ChatProxyConfig.from_env()", () => {
    const env = buildProxyEnv({ mode: "direct", apiKey: "sk-ant-abc" }, { PATH: "/usr/bin" } as NodeJS.ProcessEnv);
    expect(env.ANTHROPIC_API_KEY).toBe("sk-ant-abc");
    expect(env.ANTHROPIC_AUTH_TOKEN).toBeUndefined();
    expect(env.DECOY_ANTHROPIC_UPSTREAM_URL).toBeUndefined();
    expect(env.PATH).toBe("/usr/bin"); // base env preserved
  });

  it("Org Gateway mode: sets ANTHROPIC_AUTH_TOKEN, DECOY_ANTHROPIC_UPSTREAM_URL, and DECOY_UPSTREAM_SKIP_TLS_VERIFY -- the exact env vars chat_proxy.py's ChatProxyConfig.from_env() reads", () => {
    const creds: GatewayCredentials = {
      mode: "gateway",
      authToken: "gw-token-xyz",
      baseUrl: "https://api-eu1.aigateway.emirates.group",
      skipTlsVerify: true,
    };
    const env = buildProxyEnv(creds, {} as NodeJS.ProcessEnv);
    expect(env.ANTHROPIC_AUTH_TOKEN).toBe("gw-token-xyz");
    expect(env.DECOY_ANTHROPIC_UPSTREAM_URL).toBe("https://api-eu1.aigateway.emirates.group");
    expect(env.DECOY_UPSTREAM_SKIP_TLS_VERIFY).toBe("true");
    expect(env.ANTHROPIC_API_KEY).toBeUndefined();
  });

  it("Org Gateway mode with TLS-skip off sends the flag as the literal string 'false', not omitted", () => {
    const creds: GatewayCredentials = { mode: "gateway", authToken: "t", baseUrl: "https://gw.example.com", skipTlsVerify: false };
    const env = buildProxyEnv(creds, {} as NodeJS.ProcessEnv);
    expect(env.DECOY_UPSTREAM_SKIP_TLS_VERIFY).toBe("false");
  });
});

File: intellij-plugin/core/src/test/kotlin/com/decoy/core/CredentialConfigTest.kt
package com.decoy.core

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

/** A real in-memory implementation, not a mocking-framework double --
 * mirrors vscode-extension/src/credentialConfig.test.ts's makeFakeHost.
 * `store` doubles as the "secure storage" for persistence-across-restart
 * assertions: a fresh FakeCredentialsHost pointed at the SAME store map
 * simulates "the IDE restarted, PasswordSafe is still there". */
class FakeCredentialsHost(
    val store: MutableMap<String, String> = mutableMapOf(),
    private val choice: String? = null,
    private val gatewayToken: String? = null,
    private val gatewayUrl: String? = null,
    private val skipTlsVerify: Boolean = false,
    private val apiKey: String? = null,
) : SecretsHost {
    val infoMessages = mutableListOf<String>()
    val errorMessages = mutableListOf<String>()

    override fun getSecret(key: String): String? = store[key]
    override fun storeSecret(key: String, value: String) { store[key] = value }
    override fun deleteSecret(key: String) { store.remove(key) }
    override fun promptForApiKey(promptText: String): String? = apiKey
    override fun promptForChoice(promptText: String, choices: List<String>): String? = choice
    override fun promptForText(promptText: String, password: Boolean): String? =
        if (promptText.contains("gateway URL") || promptText.contains("base URL")) gatewayUrl else gatewayToken
    override fun confirmDangerousToggle(message: String): Boolean = skipTlsVerify
    override fun showInfo(message: String) { infoMessages.add(message) }
    override fun showError(message: String) { errorMessages.add(message) }
}

class CredentialConfigTest {

    @Test
    fun `runSetCredentials cancel at the mode choice stores nothing`() {
        val host = FakeCredentialsHost(choice = null)
        assertFalse(runSetCredentials(host))
        assertTrue(host.store.isEmpty())
    }

    @Test
    fun `runSetCredentials Direct mode stores the API key and mode flag`() {
        val host = FakeCredentialsHost(choice = MODE_LABEL_DIRECT, apiKey = "sk-ant-abc")
        assertTrue(runSetCredentials(host))
        assertEquals("sk-ant-abc", host.store[SECRET_KEY_ANTHROPIC_API_KEY])
        assertEquals("direct", host.store[SECRET_KEY_CREDENTIAL_MODE])
    }

    @Test
    fun `runSetCredentials Org Gateway mode stores token, URL, mode flag, and TLS-skip=false by default`() {
        val host = FakeCredentialsHost(
            choice = MODE_LABEL_GATEWAY,
            gatewayToken = "  gw-token-123  ",
            gatewayUrl = "https://api-eu1.aigateway.example.com",
        )
        assertTrue(runSetCredentials(host))
        assertEquals("gw-token-123", host.store[SECRET_KEY_GATEWAY_AUTH_TOKEN])
        assertEquals("https://api-eu1.aigateway.example.com", host.store[SECRET_KEY_GATEWAY_BASE_URL])
        assertEquals("false", host.store[SECRET_KEY_GATEWAY_SKIP_TLS_VERIFY])
        assertEquals("gateway", host.store[SECRET_KEY_CREDENTIAL_MODE])
        assertFalse(host.infoMessages[0].contains("gw-token-123"))
    }

    @Test
    fun `runSetCredentials Org Gateway mode with TLS-skip explicitly confirmed stores true and warns in the confirmation`() {
        val host = FakeCredentialsHost(
            choice = MODE_LABEL_GATEWAY, gatewayToken = "gw-token", gatewayUrl = "https://gw.example.com", skipTlsVerify = true,
        )
        runSetCredentials(host)
        assertEquals("true", host.store[SECRET_KEY_GATEWAY_SKIP_TLS_VERIFY])
        assertTrue(host.infoMessages[0].contains("DISABLED"))
    }

    @Test
    fun `runSetCredentials cancelling the token prompt in gateway mode stores nothing`() {
        val host = FakeCredentialsHost(choice = MODE_LABEL_GATEWAY, gatewayToken = null, gatewayUrl = "https://gw.example.com")
        assertFalse(runSetCredentials(host))
        assertTrue(host.store.isEmpty())
    }

    @Test
    fun `runSetCredentials cancelling the URL prompt in gateway mode stores nothing, not even the already-entered token`() {
        val host = FakeCredentialsHost(choice = MODE_LABEL_GATEWAY, gatewayToken = "gw-token", gatewayUrl = null)
        assertFalse(runSetCredentials(host))
        assertTrue(host.store.isEmpty())
    }

    @Test
    fun `ensureCredentials resolves stored Direct-mode credentials without prompting`() {
        val host = FakeCredentialsHost(
            store = mutableMapOf(SECRET_KEY_ANTHROPIC_API_KEY to "sk-ant-existing", SECRET_KEY_CREDENTIAL_MODE to "direct"),
        )
        val creds = ensureCredentials(host)
        assertEquals(ResolvedCredentials.Direct("sk-ant-existing"), creds)
    }

    @Test
    fun `ensureCredentials resolves stored Org Gateway credentials without prompting`() {
        val host = FakeCredentialsHost(
            store = mutableMapOf(
                SECRET_KEY_CREDENTIAL_MODE to "gateway",
                SECRET_KEY_GATEWAY_AUTH_TOKEN to "gw-token",
                SECRET_KEY_GATEWAY_BASE_URL to "https://gw.example.com",
                SECRET_KEY_GATEWAY_SKIP_TLS_VERIFY to "true",
            ),
        )
        val creds = ensureCredentials(host)
        assertEquals(ResolvedCredentials.Gateway("gw-token", "https://gw.example.com", true), creds)
    }

    @Test
    fun `credential-mode choice persists across a simulated IDE restart (fresh host, same backing store)`() {
        val sharedStore = mutableMapOf<String, String>()
        val firstSessionHost = FakeCredentialsHost(
            store = sharedStore, choice = MODE_LABEL_GATEWAY, gatewayToken = "gw-token-persisted", gatewayUrl = "https://gw.example.com",
        )
        runSetCredentials(firstSessionHost)

        // A brand-new host instance pointed at the same store simulates a
        // fresh plugin/IDE activation after a restart -- no prompt values
        // are configured, so if this had to prompt again the assertion
        // below would fail (only getSecret is exercised).
        val secondSessionHost = FakeCredentialsHost(store = sharedStore)
        val creds = ensureCredentials(secondSessionHost)
        assertEquals(ResolvedCredentials.Gateway("gw-token-persisted", "https://gw.example.com", false), creds)
    }

    @Test
    fun `ensureCredentials falls through to re-prompting if mode says gateway but credentials are incomplete`() {
        val host = FakeCredentialsHost(
            store = mutableMapOf(SECRET_KEY_CREDENTIAL_MODE to "gateway"), // token/URL missing
            choice = MODE_LABEL_GATEWAY, gatewayToken = "recovered-token", gatewayUrl = "https://gw.example.com",
        )
        val creds = ensureCredentials(host)
        assertTrue(creds is ResolvedCredentials.Gateway)
        assertEquals("recovered-token", (creds as ResolvedCredentials.Gateway).authToken)
    }

    @Test
    fun `ensureCredentials returns null if the user cancels the prompt with nothing stored yet`() {
        val host = FakeCredentialsHost(choice = null)
        assertNull(ensureCredentials(host))
    }

    @Test
    fun `clearCredentials removes all credential keys for both modes`() {
        val host = FakeCredentialsHost(
            store = mutableMapOf(
                SECRET_KEY_ANTHROPIC_API_KEY to "sk-ant",
                SECRET_KEY_GATEWAY_AUTH_TOKEN to "gw-token",
                SECRET_KEY_GATEWAY_BASE_URL to "https://gw.example.com",
                SECRET_KEY_GATEWAY_SKIP_TLS_VERIFY to "true",
                SECRET_KEY_CREDENTIAL_MODE to "gateway",
            ),
        )
        clearCredentials(host)
        assertTrue(host.store.isEmpty())
    }

    @Test
    fun `buildProxyEnv Direct mode sets ANTHROPIC_API_KEY only, matching chat_proxy py's ChatProxyConfig from_env`() {
        val env = buildProxyEnv(ResolvedCredentials.Direct("sk-ant-abc"), mapOf("PATH" to "/usr/bin"))
        assertEquals("sk-ant-abc", env["ANTHROPIC_API_KEY"])
        assertNull(env["ANTHROPIC_AUTH_TOKEN"])
        assertNull(env["DECOY_ANTHROPIC_UPSTREAM_URL"])
        assertEquals("/usr/bin", env["PATH"])
    }

    @Test
    fun `buildProxyEnv Org Gateway mode sets ANTHROPIC_AUTH_TOKEN, DECOY_ANTHROPIC_UPSTREAM_URL, and DECOY_UPSTREAM_SKIP_TLS_VERIFY`() {
        val creds = ResolvedCredentials.Gateway("gw-token-xyz", "https://api-eu1.aigateway.emirates.group", true)
        val env = buildProxyEnv(creds, emptyMap())
        assertEquals("gw-token-xyz", env["ANTHROPIC_AUTH_TOKEN"])
        assertEquals("https://api-eu1.aigateway.emirates.group", env["DECOY_ANTHROPIC_UPSTREAM_URL"])
        assertEquals("true", env["DECOY_UPSTREAM_SKIP_TLS_VERIFY"])
        assertNull(env["ANTHROPIC_API_KEY"])
    }

    @Test
    fun `buildProxyEnv Org Gateway mode with TLS-skip off sends the flag as the literal string 'false', not omitted`() {
        val creds = ResolvedCredentials.Gateway("t", "https://gw.example.com", false)
        val env = buildProxyEnv(creds, emptyMap())
        assertEquals("false", env["DECOY_UPSTREAM_SKIP_TLS_VERIFY"])
    }
}
